File: LongestSubstringKChar/LongestSubstringKChar.cpp
// LongestSubstringKChar.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <string>

int CountChar(std::string::const_iterator i, int length)
{
    int count = 1;
    for (auto ii = i+1; ii != i + length; ++ii)
    {
        bool found = false;
        for (auto jj = i; jj != ii; ++jj)
        {
            if (*ii == *jj)
            {
                found = true;
                break;
            }
        }
        if (!found)
            ++count;
    }

    return count;
}

int LongestSubstringWithKUniqueCharacters(const std::string& str, int k)
{
    int length = k;
    bool gotOne = false;
    for (size_t i = 0; i < str.size() - k; ++i)
    {
        for (size_t j = length; j + i <= str.size(); ++j)
        {
            int count = CountChar(str.cbegin()+i, static_cast<int>(j));
            if (count == k)
            {
                gotOne = true;
                length = std::max(length, static_cast<int>(j));
            }
            else if (count > k)
                break;
        }
    }

    return gotOne ? length : -1;
}

int main()
{
    while (true)
    {
        std::cout << "String (\"q\" to exit): ";
        std::string s;
        std::cin >> s;
        if (s == "q" || s == "Q")
            break;

        std::cout << "How many unique characters: ";
        int k;
        std::cin >> k;

        std::cout << "The longest substring with " << k << "Unique characters is " << LongestSubstringWithKUniqueCharacters(s, k) << std::endl;
    }
}

File: Go/LongestSubstringKChar.go
package main

import "fmt"

func GetSubstringLength(s string, index, k int) int {
	m := make(map[byte]struct{})
	for i := index; i < len(s); i++ {
		_, f := m[s[i]]
		if !f {
			if len(m) == k {
				return i - index
			} else {
				m[s[i]] = struct{}{}
			}
		}
	}

	if len(m) == k {
		return len(s) - index
	}

	return -1
}

func LongestSubstringWithKUniqueCharacters(s string, k int) int {
	longest := -1

	for i := 0; i < len(s); i++ {
		longest = max(longest, GetSubstringLength(s, i, k))
	}

	return longest
}

func main() {
	for {
		fmt.Print("Please enter the string: ")
		var str string
		fmt.Scan(&str)
		if str == "q" || str == "Q" {
			break
		}

		fmt.Print("Please enter the number k: ")
		var k int
		fmt.Scan(&k)

		fmt.Println("The longest substring with K Unique characters is ", LongestSubstringWithKUniqueCharacters(str, k))
	}
}
